File: backend/app/middleware/rate_limit.py
from fastapi import Request, HTTPException
from fastapi.responses import JSONResponse
import time
import logging
from typing import Dict, Tuple
from app.config import settings

logger = logging.getLogger(__name__)

class RateLimitMiddleware:
    """Rate limiting middleware for FastAPI"""
    
    def __init__(self, app):
        self.app = app
        self.requests: Dict[str, Tuple[int, float]] = {}
        self.cleanup_interval = 60  # Clean up every 60 seconds
        self.last_cleanup = time.time()
    
    async def __call__(self, scope, receive, send):
        if scope["type"] == "http":
            # Get client identifier
            client_id = self._get_client_id(scope)
            
            # Check rate limit
            if not self._check_rate_limit(client_id):
                # Rate limit exceeded
                response = JSONResponse(
                    status_code=429,
                    content={
                        "success": False,
                        "error": "Rate limit exceeded",
                        "code": "RATE_LIMIT_EXCEEDED",
                        "retry_after": settings.RATE_LIMIT_WINDOW,
                        "limit": settings.RATE_LIMIT_MAX_REQUESTS,
                        "window": settings.RATE_LIMIT_WINDOW
                    }
                )
                
                # Add rate limit headers
                response.headers["Retry-After"] = str(settings.RATE_LIMIT_WINDOW)
                response.headers["X-RateLimit-Limit"] = str(settings.RATE_LIMIT_MAX_REQUESTS)
                response.headers["X-RateLimit-Remaining"] = "0"
                response.headers["X-RateLimit-Reset"] = str(int(time.time() + settings.RATE_LIMIT_WINDOW))
                
                await response(scope, receive, send)
                return
            
            # Clean up old entries periodically
            self._cleanup_old_entries()
        
        await self.app(scope, receive, send)
    
    def _get_client_id(self, scope: dict) -> str:
        """Get unique client identifier"""
        # Get client IP
        client_ip = scope.get("client", ("unknown", 0))[0]
        
        # Get user agent if available
        headers = dict(scope.get("headers", []))
        user_agent = headers.get(b"user-agent", b"unknown").decode("utf-8", errors="ignore")
        
        # Create client identifier
        client_id = f"{client_ip}:{user_agent[:50]}"
        return client_id
    
    def _check_rate_limit(self, client_id: str) -> bool:
        """Check if client is within rate limit"""
        current_time = time.time()
        
        if client_id in self.requests:
            request_count, window_start = self.requests[client_id]
            
            # Check if window has expired
            if current_time - window_start > settings.RATE_LIMIT_WINDOW:
                # Reset window
                self.requests[client_id] = (1, current_time)
                return True
            
            # Check if limit exceeded
            if request_count >= settings.RATE_LIMIT_MAX_REQUESTS:
                return False
            
            # Increment request count
            self.requests[client_id] = (request_count + 1, window_start)
        else:
            # First request from this client
            self.requests[client_id] = (1, current_time)
        
        return True
    
    def _cleanup_old_entries(self):
        """Clean up expired rate limit entries"""
        current_time = time.time()
        
        # Only cleanup every cleanup_interval seconds
        if current_time - self.last_cleanup < self.cleanup_interval:
            return
        
        self.last_cleanup = current_time
        
        # Remove expired entries
        expired_clients = []
        for client_id, (_, window_start) in self.requests.items():
            if current_time - window_start > settings.RATE_LIMIT_WINDOW:
                expired_clients.append(client_id)
        
        for client_id in expired_clients:
            del self.requests[client_id]
        
        if expired_clients:
            logger.debug(f"Cleaned up {len(expired_clients)} expired rate limit entries")

# Rate limit decorator for specific endpoints
def rate_limit(max_requests: int = None, window: int = None):
    """Decorator to apply custom rate limiting to endpoints"""
    def decorator(func):
        async def wrapper(*args, **kwargs):
            # This would be implemented per-endpoint if needed
            return await func(*args, **kwargs)
        return wrapper
    return decorator

# Predefined rate limit configurations
RATE_LIMIT_CONFIGS = {
    "auth": {
        "max_requests": 5,
        "window": 900  # 15 minutes
    },
    "upload": {
        "max_requests": 10,
        "window": 60   # 1 minute
    },
    "api": {
        "max_requests": 60,
        "window": 60   # 1 minute
    },
    "public": {
        "max_requests": 100,
        "window": 60   # 1 minute
    }
}

File: lib/middleware/rateLimit.ts
import { NextRequest, NextResponse } from 'next/server'

export interface RateLimitConfig {
  windowMs: number // Time window in milliseconds
  maxRequests: number // Maximum requests per window
  keyGenerator?: (req: NextRequest) => string // Custom key generator
  skipSuccessfulRequests?: boolean // Skip rate limiting for successful requests
  skipFailedRequests?: boolean // Skip rate limiting for failed requests
  handler?: (req: NextRequest, res: NextResponse) => NextResponse // Custom handler
}

interface RateLimitEntry {
  count: number
  resetTime: number
}

class RateLimitStore {
  private store = new Map<string, RateLimitEntry>()
  private cleanupInterval: NodeJS.Timeout

  constructor() {
    // Clean up expired entries every 5 minutes
    this.cleanupInterval = setInterval(() => {
      this.cleanup()
    }, 5 * 60 * 1000)
  }

  private cleanup() {
    const now = Date.now()
    const entries = Array.from(this.store.entries())
    for (const [key, entry] of entries) {
      if (entry.resetTime < now) {
        this.store.delete(key)
      }
    }
  }

  get(key: string): RateLimitEntry | undefined {
    return this.store.get(key)
  }

  set(key: string, entry: RateLimitEntry) {
    this.store.set(key, entry)
  }

  destroy() {
    clearInterval(this.cleanupInterval)
    this.store.clear()
  }
}

// Global rate limit store
const rateLimitStore = new RateLimitStore()

export function createRateLimit(config: RateLimitConfig) {
  return function rateLimit(req: NextRequest): NextResponse | null {
    const key = config.keyGenerator ? config.keyGenerator(req) : getClientKey(req)
    const now = Date.now()
    
    // Get current rate limit entry
    let entry = rateLimitStore.get(key)
    
    if (!entry || entry.resetTime < now) {
      // Create new entry or reset existing one
      entry = {
        count: 0,
        resetTime: now + config.windowMs
      }
    }
    
    // Check if limit exceeded
    if (entry.count >= config.maxRequests) {
      const retryAfter = Math.ceil((entry.resetTime - now) / 1000)
      
      if (config.handler) {
        return config.handler(req, NextResponse.json(
          { 
            success: false, 
            error: 'Rate limit exceeded',
            code: 'RATE_LIMIT_EXCEEDED',
            retryAfter,
            limit: config.maxRequests,
            windowMs: config.windowMs
          },
          { 
            status: 429,
            headers: {
              'Retry-After': retryAfter.toString(),
              'X-RateLimit-Limit': config.maxRequests.toString(),
              'X-RateLimit-Remaining': '0',
              'X-RateLimit-Reset': new Date(entry.resetTime).toISOString()
            }
          }
        ))
      }
      
      return NextResponse.json(
        { 
          success: false, 
          error: 'Rate limit exceeded',
          code: 'RATE_LIMIT_EXCEEDED',
          retryAfter,
          limit: config.maxRequests,
          windowMs: config.windowMs
        },
        { 
          status: 429,
          headers: {
            'Retry-After': retryAfter.toString(),
            'X-RateLimit-Limit': config.maxRequests.toString(),
            'X-RateLimit-Remaining': '0',
            'X-RateLimit-Reset': new Date(entry.resetTime).toISOString()
          }
        }
      )
    }
    
    // Increment counter
    entry.count++
    rateLimitStore.set(key, entry)
    
    // Add rate limit headers to response
    const response = NextResponse.next()
    response.headers.set('X-RateLimit-Limit', config.maxRequests.toString())
    response.headers.set('X-RateLimit-Remaining', (config.maxRequests - entry.count).toString())
    response.headers.set('X-RateLimit-Reset', new Date(entry.resetTime).toISOString())
    
    return null // Continue to next middleware/handler
  }
}

function getClientKey(req: NextRequest): string {
  // Use IP address as default key
  const ip = req.headers.get('x-forwarded-for') || req.headers.get('x-real-ip') || 'unknown'
  const userAgent = req.headers.get('user-agent') || 'unknown'
  
  // Create a hash-like key from IP and user agent
  return `${ip}:${userAgent.substring(0, 50)}`
}

// Predefined rate limit configurations
export const rateLimitConfigs = {
  // Strict rate limiting for authentication endpoints
  auth: {
    windowMs: 15 * 60 * 1000, // 15 minutes
    maxRequests: 5, // 5 requests per 15 minutes
    keyGenerator: (req: NextRequest) => `auth:${getClientKey(req)}`
  },
  
  // Moderate rate limiting for API endpoints
  api: {
    windowMs: 1 * 60 * 1000, // 1 minute
    maxRequests: 60, // 60 requests per minute
    keyGenerator: (req: NextRequest) => `api:${getClientKey(req)}`
  },
  
  // Generous rate limiting for public endpoints
  public: {
    windowMs: 1 * 60 * 1000, // 1 minute
    maxRequests: 100, // 100 requests per minute
    keyGenerator: (req: NextRequest) => `public:${getClientKey(req)}`
  },
  
  // Strict rate limiting for file uploads
  upload: {
    windowMs: 1 * 60 * 1000, // 1 minute
    maxRequests: 10, // 10 uploads per minute
    keyGenerator: (req: NextRequest) => `upload:${getClientKey(req)}`
  }
}

// Helper function to apply rate limiting
export function withRateLimit(
  config: RateLimitConfig,
  handler: (req: NextRequest) => Promise<NextResponse>
) {
  return async (req: NextRequest): Promise<NextResponse> => {
    const rateLimit = createRateLimit(config)
    const rateLimitResult = rateLimit(req)
    
    if (rateLimitResult) {
      return rateLimitResult
    }
    
    return handler(req)
  }
}
